### PR TITLE
Pipe: add memory / cache hit rate / resource reference count / tsfile epoch metrics

### PR DESCRIPTION
**Pull request by VGalaxies**
## Description

As title, also fix potential NPE when marking...

Pipe metrics doc 👉 https://apache-iotdb.feishu.cn/docx/QW6edUm0KoVP7lxPXhycMLvJn7g

**NOTE:** Since there is no observed logic for the end of the `WALInsertNodeCache` lifecycle, the `PipeWALInsertNodeCacheMetrics::deregister` is currently not being called.

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [x] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/service/metric/enums/Metric.java [hunk 1]
```diff
@@ -118,6 +118,15 @@ public enum Metric {
...
   PIPE_HEARTBEAT_EVENT("pipe_heartbeat_event"),
+  PIPE_WAL_INSERT_NODE_CACHE_HIT_RATE("pipe_wal_insert_node_cache_hit_rate"),
+  PIPE_TSFILE_EPOCH_EMPTY_STATE_COUNT("pipe_tsfile_epoch_empty_state_count"),
+  PIPE_TSFILE_EPOCH_TABLET_STATE_COUNT("pipe_tsfile_epoch_tablet_state_count"),
+  PIPE_TSFILE_EPOCH_TSFILE_STATE_COUNT("pipe_tsfile_epoch_tsfile_state_count"),
+  PIPE_TSFILE_EPOCH_BOTH_STATE_COUNT("pipe_tsfile_epoch_both_state_count"),
+  PIPE_MEM_COST("pipe_mem_cost"),
+  PIPE_MEM_USAGE("pipe_mem_usage"),
+  PINNED_WAL_COUNT("pinned_wal_count"),
+  LINKED_TSFILE_COUNT("linked_tsfile_count"),
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/service/metric/enums/Metric.java:129]:
```suggestion
  PIPE_PINNED_WAL_COUNT("pipe_pinned_wal_count"),
  PIPE_LINKED_TSFILE_COUNT("pipe_linked_tsfile_count"),
```

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/service/metric/enums/Metric.java [hunk 2]
```diff
@@ -118,6 +118,15 @@ public enum Metric {
...
   PIPE_CONNECTOR_HEARTBEAT_TRANSFER("pipe_connector_heartbeat_transfer"),
   PIPE_HEARTBEAT_EVENT("pipe_heartbeat_event"),
+  PIPE_WAL_INSERT_NODE_CACHE_HIT_RATE("pipe_wal_insert_node_cache_hit_rate"),
+  PIPE_TSFILE_EPOCH_EMPTY_STATE_COUNT("pipe_tsfile_epoch_empty_state_count"),
+  PIPE_TSFILE_EPOCH_TABLET_STATE_COUNT("pipe_tsfile_epoch_tablet_state_count"),
+  PIPE_TSFILE_EPOCH_TSFILE_STATE_COUNT("pipe_tsfile_epoch_tsfile_state_count"),
+  PIPE_TSFILE_EPOCH_BOTH_STATE_COUNT("pipe_tsfile_epoch_both_state_count"),
+  PIPE_MEM_COST("pipe_mem_cost"),
+  PIPE_MEM_USAGE("pipe_mem_usage"),
+  PINNED_WAL_COUNT("pinned_wal_count"),
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/service/metric/enums/Metric.java:128]:
```suggestion
  PINNED_MEMTABLE_COUNT("pinned_memtable_count"),
```

**Review by SteveYurongSu — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.Map;
+import java.util.Objects;
+
+public class PipeWALInsertNodeCacheMetrics implements IMetricSet {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PipeWALInsertNodeCacheMetrics.class);
+
+  private AbstractMetricService metricService;
+
+  private final Map<Integer, WALInsertNodeCache> cacheMap = new HashMap<>();
```
Comment (SpriCoder) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java:45]: Why not consider ConcurrentHashMap, and try to use CAS when update

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  //////////////////////////// register & deregister (pipe integration) ////////////////////////////
+
+  public void register(@NonNull WALInsertNodeCache walInsertNodeCache, Integer dataRegionId) {
+    synchronized (this) {
+      cacheMap.putIfAbsent(dataRegionId, walInsertNodeCache);
+      if (Objects.nonNull(metricService)) {
+        createMetrics(dataRegionId);
+      }
+    }
```
Comment (SpriCoder) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java:104]: In this place, if cacheMap is concurrentHashMap,
We can use compute method here which is CAS

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+      if (!cacheMap.containsKey(dataRegionId)) {
+        LOGGER.warn(
+            "Failed to deregister wal insert node cache metrics, WALInsertNodeCache({}) does not exist",
+            dataRegionId);
+        return;
+      }
+      if (Objects.nonNull(metricService)) {
+        removeMetrics(dataRegionId);
+      }
+      cacheMap.remove(dataRegionId);
```
Comment (SpriCoder) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/metric/PipeWALInsertNodeCacheMetrics.java:119]: Same.

**Review by SpriCoder — CHANGES_REQUESTED** [review]
Excellent Job! Some concurrent suggestions PTAL~

**Review by MiniSho — APPROVED** [review]
LGTM 🚀

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**